pkg/controller/mizar/mizar-node-controller.go

```diff
@@ -0,0 +1,210 @@
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %v controller", controllerForMizarNode)
+	defer klog.Infof("Shutting down %v controller", controllerForMizarNode)
+
+	if !controller.WaitForCacheSync(controllerForMizarNode, stopCh, c.listerSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *MizarNodeController) createObj(obj interface{}) {
+	key, _ := controller.KeyFunc(obj)
+	c.queue.Add(KeyWithEventType{Key: key, EventType: EventType_Create})
+}
+
+// When an object is updated.
+func (c *MizarNodeController) updateObj(old, cur interface{}) {
+	curObj := cur.(*v1.Node)
+	oldObj := old.(*v1.Node)
+	if curObj.ResourceVersion == oldObj.ResourceVersion {
+		// Periodic resync will send update events for all known objects.
+		// Two different versions of the same object will always have different RVs.
+		return
+	}
+	if curObj.DeletionTimestamp != nil {
+		// Object is being deleted. Don't handle update anymore.
+		return
+	}
+
+	key, _ := controller.KeyFunc(curObj)
+	c.queue.Add(KeyWithEventType{Key: key, EventType: EventType_Update, ResourceVersion: curObj.ResourceVersion})
+}
+
+func (c *MizarNodeController) deleteObj(obj interface{}) {
+	key, _ := controller.KeyFunc(obj)
+	klog.Infof("%v deleted. key %s.", controllerForMizarNode, key)
+	c.queue.Add(KeyWithEventType{Key: key, EventType: EventType_Delete})
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the handler is never invoked concurrently with the same key.
+func (c *MizarNodeController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *MizarNodeController) processNextWorkItem() bool {
+	workItem, quit := c.queue.Get()
+
+	if quit {
+		return false
+	}
+
+	keyWithEventType := workItem.(KeyWithEventType)
+	key := keyWithEventType.Key
+	defer c.queue.Done(key)
+
+	err := c.handler(keyWithEventType)
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("Handle %v of key %v failed with %v", controllerForMizarNode, key, err))
+	c.queue.AddRateLimited(keyWithEventType)
+
+	return true
+}
+
+func (c *MizarNodeController) handle(keyWithEventType KeyWithEventType) error {
+	key := keyWithEventType.Key
+	eventType := keyWithEventType.EventType
+	klog.Infof("Entering handling for %v. key %s, eventType %s", controllerForMizarNode, key, eventType)
+
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished handling %v %q (%v)", controllerForMizarNode, key, time.Since(startTime))
+	}()
+
+	tenant, namespace, name, err := cache.SplitMetaTenantNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, err := c.lister.Get(name)
+	if err != nil {
+		return err
+	}
+
+	klog.V(4).Infof("Handling %v %s/%s/%s hashkey %v for event %v", controllerForMizarNode, tenant, namespace, obj.Name, obj.HashKey, eventType)
```

Comment:
the tenant and namespaces of a node are both empty, since it is cluster-scoped. Logging them will make the log harder to read